sqlx-core/src/pool/options.rs

```diff
@@ -33,6 +34,7 @@ pub struct PoolOptions<DB: Database> {
     pub(crate) max_lifetime: Option<Duration>,
     pub(crate) idle_timeout: Option<Duration>,
     pub(crate) fair: bool,
+    pub(crate) metric_observer: Option<Arc<dyn PoolMetricsObserver>>,
```

Comment:
By using `dyn` here we don't have to add a generic type+bound to the `PoolOptions` (and `Pool`, `SharedPool`, etc) - this means it's not a breaking change, though really there's no need for it to be `dyn`.

We could instead add use the generic type / static dispatch, and set a [default type][1] to a no-op impl so that users don't have to explicitly write the type. I _think_ this would be backwards compatible. I'd personally do this, but wanted to present the `dyn` version first as it is the least contentious :+1:

[1]:https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#default-generic-type-parameters-and-operator-overloading

Reply:
`dyn` is fine, I definitely don't want to leak a type parameter everywhere.

* None of the trait methods should be called in hot code paths.
* As shown above, LLVM appears to be capable of inlining dynamic function calls, presumably if it can prove that the invoked function pointer is always the same one.

Reply:
It's cool LLVM has the ability to infer and inline the same `dyn` usage, but there's no way it'll be able to do it if the `dyn` is runtime controllable. Our usage of instrumentation is typically enabled by a CLI flag to toggle it on or off, in which case we still wind up paying for it being "off" - I don't think this is a prohibitive cost, but it's an interesting observation if part of the reasoning is "LLVM can inline it".

I don't know what else you're planning to instrument with this, so I'll trust you when you say it'll never be in the hot path :+1: